docs/proposals/blocks-storage-sharding.md

```diff
@@ -0,0 +1,84 @@
+
+Given this assumption, we want to avoid sending every query to every store-gateway instance. The querier should be able to take an informed decision about the minimum subset of store-gateway instances which needs to query given a time range.
+
+The idea is to run the [`MetaFetcher`](https://github.com/thanos-io/thanos/blob/master/pkg/block/fetcher.go#L127) also within the querier, but without any sharding filter (contrary to the store-gateway). At any given point in time, the querier knows the entire list of blocks in the storage. When the querier executes the `Select()` (or `SelectSorted()`) it does:
+
+1. Compute the list of blocks by the query time range
+2. Compute the minimum list of store-gateway instances containing the required blocks (using the information from the ring)
+3. Fetch series from ingesters and the matching store-gateway instances
+4. Merge and deduplicate received series
+   - Optimization: can be skipped if the querier hits only 1 store-gateway
+
+### Query results consistency
+
+When a querier executes a query, it should guarantee that either all blocks matching the time range are queried or the query fails.
+
+However, due to the (intentional) lack of a strong coordination between queriers and store-gateways, and the ring topology which can change any time, there's no guarantee that the blocks assigned to a store-gateway shard are effectively loaded on the store-gateway itself at any given point in time.
+
+The idea is introduce a **consistency check in the querier**. When a store-gateway receives a request from the querier, the store-gateway includes in the response the list of block IDs currently loaded on the store-gateway itself. The querier can then merge the list of block IDs received from all store-gateway hit, and match it against the list of block IDs computed at the beginning of the query execution.
+
+There are three possible scenarios:
+
+1. The list match: all good
+2. All the blocks known by the querier are within the list of blocks returned by store-gateway, but the store-gateway also included blocks unknown to the querier: all good (it means the store-gateways have discovered and loaded new blocks before the querier discovered them)
+3. Some blocks known by the querier are **not** within the list of blocks returned by store-gateway: potential consistency issue
+
+We want to protect from a partial results response which may occur in the case #3. However, there are some legit cases which, if not handled, would lead to frequent false positives. Given the querier and store-gateway instances independently scan the bucket at a regular interval (to find new blocks or deleted blocks), we may be in one of the following cases:
+
+a. The querier has discovered new blocks before the store-gateway successfully discovered and loaded them
+b. The store-gateway has offloaded blocks "marked for deletion" before the querier
```

Comment:
It's rather, before other store gateway loaded new ones, right?

Also note as our current logic offloads those blocks UNTIL replacement is visible OR delay happens. (half of delete delay), so you might be fine here. (no `b` case)